extended/src/main/java/io/kubernetes/client/extended/workqueue/WorkQueue.java

```diff
@@ -0,0 +1,122 @@
+package io.kubernetes.client.extended.workqueue;
+
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * workqueue provides a simple queue that supports the following features:
+ *
+ * <ul>
+ *   <li>Fair: items processed in the order in which they are added.
+ *   <li>Stingy: a single item will not be processed multiple times concurrently, and if an item is
+ *       added multiple times before it can be processed, it will only be processed once.
+ *   <li>Multiple consumers and producers. In particular, it is allowed for an item to be reenqueued
+ *       while it is being processed.
+ *   <li>Shutdown notifications.
+ * </ul>
+ */
+public class WorkQueue<T> {
+  // queue defines the order in which we will work on items. Every
+  // element of queue should be in the dirty set and not in the
+  // processing set.
+  private LinkedList<T> queue;
```

Comment:
Let's use a `BlockingQueue` here

(https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/BlockingQueue.html)

To limit the potential memory size.

Reply:
let's use the work-queue implementation from another pull #524